src/validation.cpp

```diff
@@ -4727,8 +4726,9 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
     return nLoaded > 0;
 }
 
-void CChainState::CheckBlockIndex(const Consensus::Params& consensusParams)
+void CChainState::CheckBlockIndex(const CChainParams& chainparams)
 {
+    static bool fCheckBlockIndex = gArgs.GetBoolArg("-checkblockindex", chainparams.DefaultConsistencyChecks());
```

Comment:
In commit "Move CheckBlockIndex to validation.cpp" (340acd0692de198dbcc8a8ecdc48744d4f741c11)

Using a static local variable here isn't great because it makes makes the method harder to test and reuse. For example, it would be impossible to write a unit test that invokes this fucntion with different `fCheckBlockIndex` values. I haven't looked at all the other variables you want to extend this to, but using static locals could create real problems in other cases.

Would suggest making this a CChainState object member and initializing it in the constructor as an alternative to leaving it as a global, and in general avoiding static locals in application logic outside of specialized init code.

Also, it'd be good to clarify the goal and advantages of this PR. It's not obvious that it's better to parse config values when they are first used instead of in AppInitParameterInteraction. For example an advantage of parsing in AppInitParameterInteraction is that if the user specifies an invalid or unparseable value, they can get error feedback immediately, and code that runs later doesn't need to deal with errors.